src/Orleans/Streams/Internal/StreamConsumerExtension.cs

```diff
@@ -63,50 +67,90 @@ internal StreamSubscriptionHandleImpl<T> SetObserver<T>(GuidId subscriptionId, S
             }
         }
 
-        internal bool RemoveObserver(GuidId subscriptionId)
+        public Task OnSubscriptionChange<T>(Func<StreamSubscriptionHandle<T>, Task> onAdd, Func<string, IStreamIdentity, Guid, Task> onRemove = null)
+        {
+            this.onSubscriptionAddAction = onAdd;
+            this.onSubscriptinRemoveAction = onRemove;
+            return TaskDone.Done;
+        }
+
+        public async Task<bool> RemoveObserver(GuidId subscriptionId, StreamId streamId)
         {
             IStreamSubscriptionHandle ignore;
+            if(this.onSubscriptinRemoveAction != null)
```

Comment:
Hmm, I think this will have some unexpected side effects.  Let's leave the remove action out and revisit it in a future pass.